src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+            System.Diagnostics.Debug.Assert(filechangeCookie == VSConstants.VSCOOKIE_NIL);
+            if (filechangeCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+                if (fileChangeService != null)
+                {
+                    int hr = fileChangeService.AdviseFileChange(project.ProjectFile, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size), this, out filechangeCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && filechangeCookie != VSConstants.VSCOOKIE_NIL);
+                    _registeredProjects.Add(project, filechangeCookie);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called by reloadable projects upon close go unregister themselves
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        /// <returns></returns>
+        public async Task UnregisterProjectAsync(IReloadableProject project)
+        {
+            await _threadHandling.SwitchToUIThread();
+            
+            UnregisterProject(project);
+        }
+
+        private void UnregisterProject(IReloadableProject project)
+        {
+            uint filechangeCookie;
+            if (_registeredProjects.TryGetValue(project, out filechangeCookie))
+            {
+                // Remove watch
+                IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+                if (fileChangeService != null)
+                {
+                    int hr = fileChangeService.UnadviseFileChange(filechangeCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                }
+
+                // Always remove the watcher from our list
+                _registeredProjects.Remove(project);
+            }
+        }
+
+        /// <summary>
+        /// Adds a file change watcher on the project file.
+        /// </summary>
+        private async Task ConnectToSolutionEvents()
+        {
+            await _threadHandling.SwitchToUIThread();
+
+            if (_solutionEventsCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.AdviseSolutionEvents(this, out _solutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && _solutionEventsCookie != VSConstants.VSCOOKIE_NIL);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        private async Task DisconnectFromSolutionEvents()
+        {
+            if(_solutionEventsCookie != VSConstants.VSCOOKIE_NIL)
```

Comment:
Additionally, since this isn't done on the UI thread, if this gets called twice, we could have two simultaneous calls make it past here, resulting in two calls to `UnadviseSolutionEvents`, one with `VSConstants.VSCOOKIE_NIL`.